lib/sequence.js

```diff
@@ -1,86 +1,70 @@
 'use strict';
-const isPromise = require('is-promise');
-const autoBind = require('auto-bind');
-const AvaError = require('./ava-error');
 
 class Sequence {
-	constructor(tests, bail) {
-		if (!tests) {
-			throw new Error('Sequence items can\'t be undefined');
+	constructor(runnables, bail) {
+		if (!Array.isArray(runnables)) {
+			throw new TypeError('Expected an array of runnables');
 		}
 
-		this.results = [];
-		this.passed = true;
-		this.reason = null;
-		this.tests = tests;
+		this.runnables = runnables;
 		this.bail = bail || false;
-
-		autoBind(this);
 	}
+
 	run() {
-		const length = this.tests.length;
+		const iterator = this.runnables[Symbol.iterator]();
 
-		for (let i = 0; i < length; i++) {
-			// If last item failed and we should bail, return results and stop
-			if (this.bail && !this.passed) {
-				return this._results();
+		let activeRunnable;
+		const onBeforeExit = () => {
+			if (activeRunnable.finishDueToInactivity) {
```

Comment:
In `Test`. See https://github.com/avajs/ava/pull/1314/commits/3e3b2131ff08838e86e34e99940a0a8ec1208e30.